INTERNALS.md

```diff
@@ -0,0 +1,233 @@
+
+Sitting in front of the arena-level allocators is a thread-local cache
+("tcache").  Allocations up to some particular size are filled from the tcache
+when it is nonempty.  When it is empty, the tcache attempts to fill itself using
+the more centralized mechanism (using items from a slab if the relevant size is
+small, and a page allocation if the size is large).  Deallocated memory is put
+into the tcache if possible (and is flushed to the more central allocator if
+not).
+
+## Some key concepts
+
+### Size classes
+
+There are two relevant modules here:
+- `sc`: Actually sets the size classes (according to static and runtime
+    configuration options).
+- `sz`: Converts in between various notions of sizes (size in bytes, the size
+    class index, or a "page size" index, which is used in making requests from
+    the page-level allocator).
+
+### Extents
+
+An extent is a page-aligned, multiple-of-page-sized region of the virtual
+address space owned by jemalloc.
+
+### Extent states and purging
+
+Modern architectures perform virtual -> physical address translation, and modern
+OSs in turn allow features like overcommit; reserving virtual address space
+without backing it with physical pages.
+
+We say that a page of virtual memory is in one of four states:
+- Active: In use by the application
+- Dirty: No longer in use, but we haven't told the OS yet.
+- Muzzy: We've told the OS to reclaim the page, but asynchronously.  It may
+  never get reclaimed, and definitely won't if we touch it again.
+- Retained: We've told the OS to reclaim the page, and waited synchronously for
+  it to do so.  It will be zeroed upon next access.
+
+(Sometimes the latter three states are grouped together as "inactive").
+
+The process of taking a page from dirty -> muzzy -> retained is "purging".
+Earlier stages of purging consume more memory, but are less expensive (both in
+the sense that it's easier to put a page into that state, and that it's more
+CPU-efficient to reuse a page in an earlier state; there's less of a chance of
+soft faults).
+
+We don't always use the retained page state; sometimes we just unmap the extent.
+See the section on the page-level allocator below, for details.
+
+### Base allocators, bootstrapping
+
+The `base` module defines the lowest-level of allocation support.  It supports
+allocation-only (no freeing), and can be replaced with "extent hooks" that
+replace the OS-level page manipulation facilities.  Each arena is associated
+with a single base allocator.  Each `base_t` has a unique index, and each arena
+is associated with a single base allocator (whose index is the user-visible
+index for the arena).
+
+Base 0 and arena 0 are used for internal metadata allocations.  They're
+sometimes called "b0" and "a0" for short.
+
+### Thread-specific data
+
+The `tsd_t` struct contains all the per-thread data used by any module.  We
+abstract this out into one place for two reasons:
+1. So that it's easy for us to control the  layout
```

Comment:
Nit: two spaces here

Reply:
@davidtgoldblatt : before landing, maybe keep the revision commits, so that the increments can be checked separately?

Reply:
Will do